DataFormats/L1TGlobal/interface/GlobalObject.h

```diff
@@ -29,6 +29,8 @@ namespace l1t {
     gtMinBiasHFM0,
     gtMinBiasHFP1,
     gtMinBiasHFM1,
+    gtZDCP,
+    gtZDCM,
```

Comment:
@missirol checked that these values are stored in the `GlobalObjectMapRecord` product, which is produced by the HLT and is stored in the RAW data.

So, adding the new entries halfway through the `enum` will change the meaning of the RAW data collected with all previous releases - which does not seem like a good idea.

Can you move the new entries at the end of the `enum` ?

Reply:
Will do, should have appended in the first place - just to check, I'm guessing I want them before ObjNull at the end of the enum? so move them between gtExternal on L47 and ObjNull on L48

Reply:
I guess so, though it would be nice if @cms-sw/l1-l2 experts could comment.

Reply:
> these values are stored in the `GlobalObjectMapRecord` product

(Just adding some info for future reference.)

`GlobalObjectMapRecord` is one of the products of `L1TGlobalProducer` (L1T emulator)
https://github.com/cms-sw/cmssw/blob/19c178740257eb48367778593da55dcad08b7a4f/L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc#L204

, it and contains `GlobalObjectMap`s
https://github.com/cms-sw/cmssw/blob/19c178740257eb48367778593da55dcad08b7a4f/DataFormats/L1TGlobal/interface/GlobalObjectMapRecord.h#L71

`GlobalObjectMap` contains `L1TObjectTypeInCond`s
https://github.com/cms-sw/cmssw/blob/19c178740257eb48367778593da55dcad08b7a4f/DataFormats/L1TGlobal/interface/GlobalObjectMap.h#L124

`L1TObjectTypeInCond` corresponds to `l1t::GlobalObject`s
https://github.com/cms-sw/cmssw/blob/19c178740257eb48367778593da55dcad08b7a4f/DataFormats/L1TGlobal/interface/GlobalObjectMapFwd.h#L34

The `GlobalObjectMapRecord` product created at HLT by the L1T emulator is part of the RAW content of basically all the main streams produced at HLT. See
```
keep GlobalObjectMapRecord_hltGtStage2ObjectMap_*_*
```
in most OutputModules of any recent HLT menu ([example](https://raw.githubusercontent.com/cms-sw/cmssw/19c178740257eb48367778593da55dcad08b7a4f/HLTrigger/Configuration/test/OnLine_HLT_HIon.py)).

Reply:
> I guess so, though it would be nice if https://github.com/orgs/cms-sw/teams/l1-l2 experts could comment.

Could @cms-sw/l1-l2 experts comment ? Changes to RAW data formats should be reviewed and tested carefully, clearly.

Reply:
@fwyzard @missirol Our thinking on this was that the ZDC input into the global trigger is essentially no different from many "sum" based triggers already present in the global trigger, and so the plan was essentially to just use it as one. Same dataformat, same parser in the global trigger, whatever. Unfortunately, my expertise ends where HLT begins. Is HLT going to be unable to handle an added sum type at the end of the ET sum format? I am willing to assist in any tests needed with the format change, but I don't know what those are at the moment. Are there specific things that should be tried? To the best of my knowledge right now, there hasn't been significant discussion right now within GT about where and how ZDC sums will end up in the object mapping for HLT, so I want to also turn the question around on the HLT experts a bit and ask what you need from us to include these ZDC sums.

Reply:
@aloeliger I let @missirol comment on the _use_ of the L1 energy sums at HLT.

The concern in [my comment](https://github.com/cms-sw/cmssw/pull/42634#discussion_r1302636113) above was that altering the values corresponding to the entries in the `GlobalObject` would alter the meaning of the RAW data (intended as the data tier stored on tape, not as the FED RAW data coming from the detctors) already collected.

Moving the new entries at the end should address most of the concerns.

Reply:
@fwyzard Okay, that's fine, I completely agree with that.

Reply:
> I let @missirol comment on the use of the L1 energy sums at HLT.

I don't think I can give a comprehensive answer on the fly, but the first place to look at is the plugin [`HLTL1TSeed`](https://github.com/cms-sw/cmssw/blob/1ad82ddd69bab05ef34c42f91ebcb931b0e9d2eb/HLTrigger/HLTfilters/plugins/HLTL1TSeed.cc). This is the plugin used in the HLT menus (at least, the Phase-1 menus) to filter on L1T decisions, and add to the event the L1T objects responsible for firing the selected L1T algos. One can see how types like (for example) [`l1t::gtETMHF`](https://github.com/cms-sw/cmssw/blob/1ad82ddd69bab05ef34c42f91ebcb931b0e9d2eb/HLTrigger/HLTfilters/plugins/HLTL1TSeed.cc#L750C18-L750C30) are used in that plugin. I don't know yet if it is strictly necessary to update `HLTL1TSeed` in light of this PR. I have started to test this PR locally, and hopefully I can give better feedback by the end of tomorrow. In any case, I would say that `HLTL1TSeed` should be updated in a follow-up PR by HLT, if needed.

Reply:
Fwiw, a similar update was done a while ago in https://github.com/cms-sw/cmssw/pull/36951, after data formats for L1T "muon showers" were introduced before the start of Run 3.

Reply:
> @fwyzard @missirol Our thinking on this was that the ZDC input into the global trigger is essentially no different from many "sum" based triggers already present in the global trigger, and so the plan was essentially to just use it as one. Same dataformat, same parser in the global trigger, whatever. Unfortunately, my expertise ends where HLT begins. Is HLT going to be unable to handle an added sum type at the end of the ET sum format? I am willing to assist in any tests needed with the format change, but I don't know what those are at the moment. Are there specific things that should be tried? To the best of my knowledge right now, there hasn't been significant discussion right now within GT about where and how ZDC sums will end up in the object mapping for HLT, so I want to also turn the question around on the HLT experts a bit and ask what you need from us to include these ZDC sums.

How are these sums used? If they will become pat of the global sums, then there is no technical changes (just the content of the variables changes). If the ZDC sums are valid and to be looked at on their own by L1T/HLT, then they should be in a separate object(collection) pulled out in HLTL1TSeed module, possibly with a different C++ type but for sure a separate object(collection). Please do not start to modify or append to existing collections, that just makes it harder and less transparent.
Note also the difference between no-collection and empty-collection. No-collection would express that ZDC was not in. Empty collection would express that ZDC did not record anything of use. The two should not be mixed.

Reply:
What is the L1T etsum type you want to append it to? I presume each element of that object collection would include a self-identifying entry? Ie, there is no implicit assumption of any ordering?

Reply:
I believe ordering is generally an issue - hence the request to append. We have in general moved all ZDCP and ZDCM additions to type enum's to the end of the enum. An example is in the EtSum header here:

https://github.com/cfmcginn/cmssw/blob/l1tZDCEmulator_CMSSW13_3_X/DataFormats/L1Trigger/interface/EtSum.h#L55

Reply:
Hmm, so how does that work? An instance of the EtSum class holds one specific Etsum value with its type/meaning identified by the type enum as class member. Then, a BXVector is created for the different bunch crossing assignments of that specific EtSum. Now, for a different EtSum type, a separate BXVector is created?

Reply:
@Martin-Grunewald ZDCs output could be a BX vector of EtSums, with each BX of the vector holding two sums, one of the plus type, and one of the minus type, or it could be two BX vectors one for positive type sums, and one for minus type sums. In it's current incarnation, the first formulation is used.

Reply:
Sorry, I do not understand about two EtSums in one BXVectorelement:
```
class EtSum : public L1Candidate
```
and 
```
typedef BXVector<EtSum> EtSumBxCollection;
```
so an element of a BXVector is an EtSum which can contain only ONE sum?

Reply:
The "appending" here is appending the new ZDC types to the end of the enum declaration to prevent any change in what previous enum entries mean in RAW. 

Right now, to the best of my knowledge the sums, as given to GT are not appended to any object map that GT puts out, they are simply re-using sum parsing code to prevent rewriting of code to handle sums that are functionally the same as every other sum handed over to GT, with the distinction being they are now coming from ZDC.

How these ZDC sums are handed to and parsed by HLT is up to HLT, but from the L1 perspective there is no practical difference between what ZDC hands to GT, and anything like MET or HT. I guess that, in theory, they could exist in a vacuum as data taken without these other sums, but that is the extent of the difference (at least to my eyes).

Reply:
> Sorry, I do not understand about two EtSums in one BXVectorelement:
> 
> ```
> class EtSum : public L1Candidate
> ```
> 
> and
> 
> ```
> typedef BXVector<EtSum> EtSumBxCollection;
> ```
> 
> so an element of a BXVector is an EtSum which can contain only ONE sum?

I believe a BX vector can contain multiple elements per BX.

Reply:
Yes, from a quick look at the code, `BXVector<T>(size, first, last)` will create a `BXVector` object with `size` elements of type `T` for each BX.

Reply:
> > Right now, to the best of my knowledge the sums, as given to GT are not appended to any object map that GT puts out, they are simply re-using sum parsing code to prevent rewriting of code to handle sums that are functionally the same as every other sum handed over to GT, with the distinction being they are now coming from ZDC.
> 
> Do you mean that a trigger like `L1_ETM150` does not mark the L1T ETM sum in the object map ?
> 
> I don't think I ever checked this.

No, it should? I guess I don't know how HLT maps the sums. My point was that internal to the uGT emulator, we are using whatever parser the sums use to figure out if the ZDC sums match ZDC trigger conditions. I don't know to what extent these then get added into the object map, if at all.

Reply:
> I don't think I can give a comprehensive answer on the fly, but the first place to look at is the plugin [`HLTL1TSeed`](https://github.com/cms-sw/cmssw/blob/1ad82ddd69bab05ef34c42f91ebcb931b0e9d2eb/HLTrigger/HLTfilters/plugins/HLTL1TSeed.cc). This is the plugin used in the HLT menus (at least, the Phase-1 menus) to filter on L1T decisions, and add to the event the L1T objects responsible for firing the selected L1T algos. One can see how types like (for example) [`l1t::gtETMHF`](https://github.com/cms-sw/cmssw/blob/1ad82ddd69bab05ef34c42f91ebcb931b0e9d2eb/HLTrigger/HLTfilters/plugins/HLTL1TSeed.cc#L750C18-L750C30) are used in that plugin. I don't know yet if it is strictly necessary to update `HLTL1TSeed` in light of this PR. I have started to test this PR locally, and hopefully I can give better feedback by the end of tomorrow. In any case, I would say that `HLTL1TSeed` should be updated in a follow-up PR by HLT, if needed.

Fwiw, I'm starting to test this PR with https://github.com/missirol/cmssw/commit/47b8866e28660a08d5a5b040ebfcbea5472f7676, which should contain the minimal changes to make HLT aware of the new L1T data formats (the fact that the underlying data format, i.e. `l1t::EtSum`, is not a new one, makes the changes on the HLT side less extensive compared to https://github.com/cms-sw/cmssw/pull/36951, as far as I understand).

Reply:
> So, adding the new entries halfway through the enum will change the meaning of the RAW data collected with all previous releases - which does not seem like a good idea.

For the record: related to this point (but unrelated to this PR), it looks like this kind of mistake was done in https://github.com/cms-sw/cmssw/pull/36018.

Reply:
Oh.

So, any Run-3 release will misinterpret the object map reading data from Run-2 ?

Reply:
That's what I would conclude.. should we open an issue to discuss how to, at least, avoid this in the future ?

Two possibilities that come to mind are (1) ask core-sw to review PRs to packages that include data formats of the RAW data tier, and (2) see if we can build on unit tests such as https://github.com/cms-sw/cmssw/pull/41565 to catch this kind of non-backward-compatible change.

Reply:
Shall we bring this up at the Core Software meeting today ?

Reply:
Yes, I think it would be useful. I'll open an issue anyway, so we have a starting point to discuss this.

Reply:
The issue is briefly summarised in https://github.com/cms-sw/cmssw/issues/42719.